lib/backend/etcdbk/etcd.go

```diff
@@ -649,6 +649,85 @@ func (b *EtcdBackend) NewWatcher(ctx context.Context, watch backend.Watch) (back
 	return b.buf.NewWatcher(ctx, watch)
 }
 
+func (b *EtcdBackend) Items(ctx context.Context, params backend.IterateParams) iter.Seq2[backend.Item, error] {
+	if params.StartKey.IsZero() {
+		err := trace.BadParameter("missing parameter startKey")
+		return func(yield func(backend.Item, error) bool) { yield(backend.Item{}, err) }
+	}
+	if params.EndKey.IsZero() {
+		err := trace.BadParameter("missing parameter endKey")
+		return func(yield func(backend.Item, error) bool) { yield(backend.Item{}, err) }
+	}
+
+	sort := clientv3.SortAscend
+	if params.Descending {
+		sort = clientv3.SortDescend
+	}
+
+	const defaultPageSize = 1000
+	return func(yield func(backend.Item, error) bool) {
+		inclusiveStartKey := b.prependPrefix(params.StartKey)
+		count := 0
+
+		pageSize := defaultPageSize
+		for {
+			if params.Limit > backend.NoLimit {
+				pageSize = min(params.Limit-count, defaultPageSize)
+			}
+
+			start := b.clock.Now()
+			re, err := b.clients.Next().Get(ctx, inclusiveStartKey,
+				// etcd's range query includes the start point and excludes the end point,
+				// but Backend.GetRange is supposed to be inclusive at both ends, so we
+				// query until the very next key in lexicographic order (i.e., the same key
+				// followed by a 0 byte)
+				clientv3.WithRange(b.prependPrefix(params.EndKey)+"\x00"),
+				clientv3.WithSort(clientv3.SortByKey, sort),
+				clientv3.WithLimit(int64(pageSize)),
+			)
+			batchReadLatencies.Observe(time.Since(start).Seconds())
+			batchReadRequests.Inc()
+			if err := convertErr(err); err != nil {
+				if !yield(backend.Item{}, trace.Wrap(err)) {
+					return
+				}
+			}
+
+			if len(re.Kvs) == 0 {
+				return
+			}
+
+			for _, kv := range re.Kvs {
+				value, err := unmarshal(kv.Value)
+				if err != nil {
+					if !yield(backend.Item{}, trace.Wrap(err)) {
+						return
+					}
+				}
+
+				if !yield(backend.Item{
+					Key:      b.trimPrefix(kv.Key),
+					Value:    value,
+					Revision: toBackendRevision(kv.ModRevision),
+				}, nil) {
+					return
+				}
+
+				count++
+				if params.Limit != backend.NoLimit && count >= params.Limit {
+					return
+				}
+			}
+
+			if params.Descending {
+				inclusiveStartKey = backend.RangeEnd(backend.KeyFromString(string(re.Kvs[0].Key))).String()
+			} else {
+				inclusiveStartKey = backend.RangeEnd(backend.KeyFromString(string(re.Kvs[len(re.Kvs)-1].Key))).String()
+			}
```

Comment:
Shouldn't we be updating the end key if we're descending? The test suite definitely needs a test with 1001+ items to check that pagination is working correctly 😬